### ISSUE TITLE
Best way to create order from plugin?

### ISSUE THREAD

**Issue by taenzerme** [pretix/pretix#936]
Hey there,

I'm trying to build a one page checkin plugin to allow to create orders in the backend easily.

Would the right way be to use pretix.base.services.orders.perform_order() or is there a better approach? If so, what is the recommended way to add positions to the order? Should I get a new cart/session id first and then use add_items_to_cart() for this?

Would it be easier to use the REST API for this?

Thanks
Sebastian

**Comment by raphaelm** [comment]
Hi,

> I'm trying to build a one page checkin plugin to allow to create orders in the backend easily.

Please be aware that while it is probably possible to do this, it *will* be incompatible to many pretix plugins, such as some payment providers, all plugins that add additional steps or fields to the checkout process, etc. You probably also end up being incompatible to some parts of pretix (like custom taxation rules) that are tightly interconnected with the shop frontend.

> Would the right way be to use pretix.base.services.orders.perform_order() or is there a better approach? If so, what is the recommended way to add positions to the order? Should I get a new cart/session id first and then use add_items_to_cart() for this?

Yes, if it is feasible for you to go through ``CartManager`` first and then through ``perform_order``, this will be best, as it performs all necessary steps such as quota checking, time frame checking, product validation, cart size validation, add-on validation, price calculation, fee calculation, e-mail dispatching, etc. Otherwise, you'd need to do all of that yourself. You still need to do a lot of things yourself, though, such as payment handling, question validation, etc…

> Would it be easier to use the REST API for this?

No. If you look at the warning block [here](https://docs.pretix.eu/en/latest/api/resources/orders.html#post--api-v1-organizers-(organizer)-events-(event)-orders-), it should become clear that the API is not intended for this usecase. You might also be interested in the thread at #599.

**Comment by taenzerme** [comment]
Sorry for reopening this one. Thanks for the response and your warning. After considering all this I decided to speed up ticket booking during checkin we simply create a plugin that loads the checkout frontend in an iframe like this:

`<iframe src="/{{request.organizer.slug}}/{{request.event.slug}}" frameborder="0" style="margin-top: 50px; width: 80vw;height: 100vh;position: relative; padding: 0;"></iframe>`

For this to work I modified the get() method for the plugins index view to allow the unsafe inline css temporarily. In the end this works fine for us, but requires to modify settings.py to allow iframes from the same origin:

`X_FRAME_OPTIONS = 'SAMEORIGIN'`

and then modify order.html quick & dirty to allow direct access to the new order:

```
        {% blocktrans trimmed with code=order.code event=event.slug %}
            Your order: <a target="_parent" href="/control/event/{{organizer}}/{{ event }}/orders/{{code}}">{{ code }}</a>
        {% endblocktrans %}
```

Is there a way to influence the checkout page somehow from my plugin or would it be a reasonable request to include the direct link like that to the order in the backend in the pretix core in order.html?
